src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,227 @@
+                                timestampCache[path] = time;
+                                Log($"Path '{path}' had live timestamp '{time}'.");
+                            }
+                            else
+                            {
+                                Log($"Path '{path}' did not exist.");
+                            }
+                        }
+                        else
+                        {
+                            Log($"Path '{path}' had cached timestamp '{time}'.");
+                        }
+
+                        if (latestTime < time)
+                        {
+                            latestTime = time;
+                        }
+                    }
+
+                    return latestTime;
+                }
+
+                if (_projectItemsSchema == null)
+                {
+                    Log($"Skipping because '{typeof(IProjectItemSchemaService).FullName}' component could not be found.");
+                    Report.IfNotPresent(_projectItemsSchema);
+                    return false;
+                }
+
+                if (!string.IsNullOrEmpty(project.GetPropertyValue(DisableFastUpToDateCheckProperty)))
+                {
+                    Log($"Disabled because the '{DisableFastUpToDateCheckProperty}' property is set to a non-empty value.");
+                    return false;
+                }
+
+                var inputs = new List<string>();
+
+                // add the project file
+                if (!string.IsNullOrEmpty(project.FullPath))
+                {
+                    inputs.Add(project.FullPath);
+                    Log($"Adding input project file {project.FullPath}.");
+                }
+
+                // add all imports except .user file
+                var userFilePath = project.FullPath + ".user";
+                var imports = project.Imports
+                    .Select(i => i.ImportedProject.FullPath)
+                    .Where(p => !string.IsNullOrEmpty(p) && !string.Equals(p, userFilePath, StringComparison.OrdinalIgnoreCase));
+
+                foreach (var import in imports)
+                {
+                    Log($"Adding input project import {import}.");
+                }
+
+                inputs.AddRange(imports);
+            
+                // add all project items (generally items seen in solution explorer) that are not excluded from UpToDate check
+                // Skip items that are marked as excluded from build.
+                var projectItemSchemaValue = (await _projectItemsSchema.GetSchemaAsync(cancellationToken)).Value;
+                var itemTypes = projectItemSchemaValue
+                    .GetKnownItemTypes()
+                    .Select(name => projectItemSchemaValue.GetItemType(name))
+                    .Where(item => item != null && item.UpToDateCheckInput && !string.Equals(item.Name, "None", StringComparison.OrdinalIgnoreCase));
```

Comment:
Yes, None should be handled in schema.

---
In reply to: [117561995](https://github.com/dotnet/project-system/pull/2241#discussion_r117561995) [](ancestors = 117561995)